### DISCUSSION TITLE
Timeline tab in grafana

### DISCUSSION THREAD

**Discussion by kimmilde** [teslamate-org/teslamate#1594] [Ideas]
Is there a possibility to get a timeline tab in grafana, like teslafi/teslalogger have? I migrated from teslafi and thats the tab I miss most

**Comment by DrMichael** [comment]
Well, teslalogger is using an own panel for that. The data would be there. 
![image](https://user-images.githubusercontent.com/8087682/117477421-47b89d80-af5e-11eb-9e42-fc1acbef777b.png)

**Reply by DrMichael** [reply]
Although, I must say, it looks really nice in teslalogger...

**Comment by kimmilde** [comment]
Yes, I agree. It gives you a very good overview for the day if it includes sleep, park, drive and charges. I would also like to have a symbol for preheating/battery heater when in park.

**Comment by DrMichael** [comment]
I worked a bit on it further...
![image](https://user-images.githubusercontent.com/8087682/117538438-7a1dd580-b006-11eb-8c83-1017c04d1891.png)

I included the dashboard. Still missing all thoese bells and whistles line km/mi, C/F, ideal/rated...

@adriankumpf what do you think should we integrate the [teslalogger-timeline-panel](https://github.com/bassmaster187/TeslaLogger/tree/master/TeslaLogger/GrafanaPlugins/teslalogger-timeline-panel/dist)?

**Reply by eliluong** [reply]
Would love this. Some of teslalogger's panels look really nice to have. I'm looking at their grafana query code, and looked at teslamate's query code for the Drives panel. But one of them uses influxdb and the other uses postgres so it is not easy for me to understand what is going on and convert between the two. Are you developing in teslamate's grafana?

**Reply by adriankumpf** [reply]
I think the author of this plugin publishes only the [minified source code](https://github.com/bassmaster187/TeslaLogger/blob/master/TeslaLogger/GrafanaPlugins/teslalogger-timeline-panel/dist/module.js) for a reason, instead of disclosing the actual source.

Since it is therefore not easily possible to make adjustments for TeslaMate, I don't think an official integration makes sense at the moment.

**Comment by DrMichael** [comment]
I had a look at the teslalogger-timeline-plugin and it is, what it says: a timeline plugin for teslalogger: paths are hardcoded, field names in te query are fixed., etc.

So, the below timeline looks really cute, but cannot / should not be done with that plugin in teslamate. If there is more interest in the above simpler timeline, I would go and finish it...

![timeline](https://user-images.githubusercontent.com/8087682/117675816-344f4180-b1ad-11eb-99d7-a6f801cdd212.png)

**Reply by kimmilde** [reply]
Yes, please. I would like the simpler timeline!

**Comment by DrMichael** [comment]
Ok, now here. Should work for miles and Fahrenheit... I am not sure about the kWh with miles set though. Please test.

@ToniA How did you avoid in https://github.com/adriankumpf/teslamate/pull/1439, that, when changing km to mi e.g., the miles fields flip to the end?

Dashboard: [timeline.zip](https://github.com/adriankumpf/teslamate/files/6459867/timeline.zip)

**Reply by ToniA** [reply]
In that case it was about adding the missing mile fields into indexByName. Check out the source code, there are not that many lines in that PR.

**Reply by eliluong** [reply]
Looks slick, thanks for posting! I'm not sure how to change km to mi, and I don't have colored icons unfortunately, probably because running on Win7.

We should have simple timeline for quick overview, which is yours, and complex timeline, like teslalogger :)

![image](https://user-images.githubusercontent.com/15806938/117872935-1ab50500-b254-11eb-9f5c-c8c20efb1fd7.png)

**Reply by DrMichael** [reply]
> Looks slick, thanks for posting! I'm not sure how to change km to mi, 

In the teslamate's settings...

> and I don't have colored icons unfortunately, probably because running on Win7.

That is definitely a font issue...

**Reply by DrMichael** [reply]
> In that case it was about adding the missing mile fields into indexByName. Check out the source code, there are not that many lines in that PR.

Ah, got that. Is there a way to define that in the Grafana GUI? Or only in the json?

**Reply by eliluong** [reply]
> In the teslamate's settings...

Oh, I see. I must not be loading it the way you intended. I had to log in to Grafana and create a new dashboard that was not attached to the set of standard Teslamate dashboards, so I don't have the usual settings there.

**Reply by DrMichael** [reply]
> > In the teslamate's settings...
> 
> Oh, I see. I must not be loading it the way you intended. I had to log in to Grafana and create a new dashboard that was not attached to the set of standard Teslamate dashboards, so I don't have the usual settings there.

Why not? You connect to the teslamate database and in that are the settings.

**Comment by DrMichael** [comment]
So, new version. I am still unclear, whether `efficiency` in `cars` is in Wh/km or Wh/mi... I have assumed always in Wh/km.

[timeline.zip](https://github.com/adriankumpf/teslamate/files/6465219/timeline.zip)

**Reply by adriankumpf** [reply]
`Wh/km` is indeed correct.

**Reply by DrMichael** [reply]
Ok. Should I make a PR out of this?

**Reply by adriankumpf** [reply]
Sure 🚀

**Reply by DrMichael** [reply]
Here we go: https://github.com/adriankumpf/teslamate/pull/1621

That is just the dashboard. Probably there need to be other references to be added, e.g. within teslamate...

**Reply by cwanja** [reply]
This looks great @DrMichael, thanks for creating it. I am assuming some of the later columns would get cleaned up? Such as:

![Screen Shot 2021-06-02 at 2 22 15 PM](https://user-images.githubusercontent.com/54645561/120540142-f3aea680-c3ad-11eb-83ef-8a165e84ca14.png)

**Reply by jun3280net** [reply]
@DrMichael  @adriankumpf  thanks for this dashboard.  This is very useful information! Any chance the Rated Range Before and After could be added?  For completeness, could Sleep be added as a state as well?  Thanks for the great work!

**Comment by DrMichael** [comment]
Sure. As soon as Grafana has fixed https://github.com/grafana/grafana/issues/24092

**Comment by DrMichael** [comment]
We took out the "missing" lines, because the previous drive does not necessarily have the id+1. I really enjoyed those lines...

If anyone could figure out, how to access in SQL the previous drive, that'd be great! @Dulanic perhaps? :-)

**Reply by Dulanic** [reply]
Looking I guessss I can do that :) I'll take a look.

**Reply by Dulanic** [reply]
I wish I had more time to work on this, but this should work.... it recreates a id based on time which i called..... time_id! :) You can adapt it to what you're doing.

```
  SELECT *
  FROM (
        SELECT row_number() over(ORDER BY start_date) AS time_id,
               *
          FROM drives
       ) d1
  LEFT JOIN (
        SELECT row_number() over(ORDER BY start_date) AS time_id,
               *
          FROM drives
       ) d2
    ON d2.time_id = d1.time_id +1
```

**Reply by Dulanic** [reply]
Oh also, to avoid all the issues around F/C and kM/Mi... try joining the settings table like so...
```
SELECT convert_celsius(outside_temp_avg, unit_of_temperature::text) AS outside_temp_avg,
       convert_km(speed_max, unit_of_length::text) AS speed_max,
       convert_km(start_ideal_range_km, unit_of_length::text) AS start_ideal_range,
       convert_km(end_ideal_range_km, unit_of_length::text) AS end_ideal_range

  FROM drives a
  JOIN settings b
    ON a.car_id = b.id
```

**Reply by DrMichael** [reply]
Shouldn't it be `d2.time_id = d1.time_id+1`?

**Reply by DrMichael** [reply]
> I wish I had more time to work on this, but this should work.... it recreates a id based on time which i called..... time_id! :) You can adapt it to what you're doing.

But how to apply it to this? The `T1.id + 1 = T2.id` is the part to replace...

```
SELECT
	T1.end_date AS "Start",
	T2.start_date AS "End",
	ROUND(EXTRACT(EPOCH FROM T2.start_date)) * 1000 - 1 AS start_date_ts,
	ROUND(EXTRACT(EPOCH FROM T2.start_date))*1000 - 1 AS end_date_ts,
	'❓ Missing' AS "Action",
	EXTRACT(EPOCH FROM T2.start_date - T1.end_date)/60 AS "Duration",
	CASE WHEN T1.end_geofence_id IS NULL THEN CONCAT('new?lat=', TP1.latitude, '&lng=', TP1.longitude)
		WHEN T1.end_geofence_id IS NOT NULL THEN CONCAT(T1.end_geofence_id, '/edit')
	END AS start_path,
	CASE WHEN T2.start_geofence_id IS NULL THEN CONCAT('new?lat=', TP2.latitude, '&lng=', TP2.longitude)
		WHEN T2.start_geofence_id IS NOT NULL THEN CONCAT(T2.start_geofence_id, '/edit')
	END AS end_path,
	COALESCE(start_geofence.name, CONCAT_WS(', ', COALESCE(start_address.name, nullif(CONCAT_WS(' ', start_address.road, start_address.house_number), '')), start_address.city)) AS "Start Address",
	COALESCE(end_geofence.name, CONCAT_WS(', ', COALESCE(end_address.name, nullif(CONCAT_WS(' ', end_address.road, end_address.house_number), '')), end_address.city)) AS "End Address",
	convert_km(TP1.odometer::INTEGER, '$length_unit') AS odometer_$length_unit,
	convert_km((TP2.odometer - TP1.odometer)::INTEGER, '$length_unit') AS distance_$length_unit,
	convert_km(((TP2.ideal_battery_range_km + (TP2.odometer - TP1.odometer)) - TP1.ideal_battery_range_km)::INTEGER, '$length_unit')::INTEGER * efficiency AS "kWh",
	convert_km(((TP2.ideal_battery_range_km + (TP2.odometer - TP1.odometer)) - TP1.ideal_battery_range_km)::INTEGER, '$length_unit') AS range_diff_$length_unit,
	TP2.battery_level AS "SoC",
	TP2.battery_level-TP1.battery_level AS "SoC Diff",
	T1.outside_temp_avg AS outside_temp_avg_$temp_unit
FROM
	d1 AS T1
	LEFT OUTER JOIN d2 AS T2 on **T1.id + 1 = T2.id**
	LEFT OUTER JOIN positions AS TP1 on T1.end_position_id = TP1.id
	LEFT OUTER JOIN positions AS TP2 on T2.start_position_id = TP2.id
	LEFT JOIN addresses start_address ON T1.end_address_id = start_address.id
	LEFT JOIN addresses end_address ON T2.start_address_id = end_address.id
	LEFT JOIN geofences start_geofence ON T1.end_geofence_id = start_geofence.id
	LEFT JOIN geofences end_geofence ON T2.start_geofence_id = end_geofence.id
	JOIN cars ON cars.id = T2.car_id
WHERE
	$__timeFilter(T1.end_date)
	AND T1.car_id=$car_id 
	AND T2.car_id=$car_id 
	AND TP2.car_id=$car_id 
	AND TP1.car_id=$car_id 
	AND TP2.odometer - TP1.odometer > 0.5
	AND T1.end_address_id <> T2.start_address_id AND T1.end_geofence_id <> T2.start_geofence_id
	AND '❓ Missing' in ($action_filter)
```

**Reply by Dulanic** [reply]
I think I missed part of what you were discussing. I'll take another look.

**Reply by Dulanic** [reply]
Can you explain what is expected? I see some showing under both? What is missing supposed to be?

![image](https://user-images.githubusercontent.com/2948206/120688211-709e5680-c468-11eb-957a-59f9e6a101c7.png)

**Reply by Dulanic** [reply]
Does this make more sense? It;s same idea, but it shows it as T1 and T2?

```
SELECT
	T1.end_date AS "Start",
	T2.start_date AS "End",
	ROUND(EXTRACT(EPOCH FROM T2.start_date)) * 1000 - 1 AS start_date_ts,
	ROUND(EXTRACT(EPOCH FROM T2.start_date))*1000 - 1 AS end_date_ts,
	'❓ Missing' AS "Action",
	EXTRACT(EPOCH FROM T2.start_date - T1.end_date)/60 AS "Duration",
	CASE WHEN T1.end_geofence_id IS NULL THEN CONCAT('new?lat=', TP1.latitude, '&lng=', TP1.longitude)
		WHEN T1.end_geofence_id IS NOT NULL THEN CONCAT(T1.end_geofence_id, '/edit')
	END AS start_path,
	CASE WHEN T2.start_geofence_id IS NULL THEN CONCAT('new?lat=', TP2.latitude, '&lng=', TP2.longitude)
		WHEN T2.start_geofence_id IS NOT NULL THEN CONCAT(T2.start_geofence_id, '/edit')
	END AS end_path,
	COALESCE(start_geofence.name, CONCAT_WS(', ', COALESCE(start_address.name, nullif(CONCAT_WS(' ', start_address.road, start_address.house_number), '')), start_address.city)) AS "Start Address",
	COALESCE(end_geofence.name, CONCAT_WS(', ', COALESCE(end_address.name, nullif(CONCAT_WS(' ', end_address.road, end_address.house_number), '')), end_address.city)) AS "End Address",
	TP1.odometer AS "Odometer",
	(TP2.odometer - TP1.odometer) AS "Distance",
	(((TP2.ideal_battery_range_km + (TP2.odometer - TP1.odometer)) - TP1.ideal_battery_range_km) * efficiency) AS "kWh",
	(((TP2.ideal_battery_range_km + (TP2.odometer - TP1.odometer)) - TP1.ideal_battery_range_km)) AS "Range Difference",
	TP1.battery_level AS "SoC",
	TP2.battery_level-TP1.battery_level AS "SoC Difference",
	T1.outside_temp_avg AS "Temperature"
FROM (SELECT row_number() over(ORDER BY start_date) AS time_id, * FROM drives) AS T1
	LEFT OUTER JOIN (SELECT row_number() over(ORDER BY start_date) AS time_id, * FROM drives ) AS T2 on T1.time_id + 1 = T2.time_id
	LEFT OUTER JOIN positions AS TP1 on T1.end_position_id = TP1.id
	LEFT OUTER JOIN positions AS TP2 on T2.start_position_id = TP2.id
	LEFT JOIN addresses start_address ON T1.end_address_id = start_address.id
	LEFT JOIN addresses end_address ON T2.start_address_id = end_address.id
	LEFT JOIN geofences start_geofence ON T1.end_geofence_id = start_geofence.id
	LEFT JOIN geofences end_geofence ON T2.start_geofence_id = end_geofence.id
	JOIN cars ON cars.id = T2.car_id
WHERE
	T1.end_date BETWEEN '2021-03-06T02:12:59.126Z' AND '2021-06-04T01:12:59.126Z'
	AND T1.car_id='1' 
	AND T2.car_id='1' 
	AND TP2.car_id='1' 
	AND TP1.car_id='1' 
	AND TP2.odometer - TP1.odometer > 0.5
	AND T1.end_address_id <> T2.start_address_id AND T1.end_geofence_id <> T2.start_geofence_id AND T1.end_position_id <> T2.start_position_id
```

**Reply by DrMichael** [reply]
> Can you explain what is expected? I see some showing under both? What is missing supposed to be?

Well, capture all drive pairs, where the end position etc. of the first is not equal to that of the second... So, something is missing there.

**Comment by DrMichael** [comment]
Right, that does the job, see attached. Still not sure, which columns really make sense...
[timeline.zip](https://github.com/adriankumpf/teslamate/files/6598171/timeline.zip)

(I saw @adriankumpf had another approach in the "parking" part...)

**Comment by LelandSindt** [comment]
Would it be possible [and reasonable to ask] that a column for  `Time Asleep` be added, and include the time (similar to `Duration`) and percentage of the `Duration` that the car is asleep... 

... as this really only applies to `Parking` events... I assume this column would be Zeros for all other events...

**Reply by jun3280net** [reply]
Thanks @DrMichael  and @Dulanic for the continued updates.  Similar to what @LelandSindt  is requesting, would it be possible to leverage the STATES data (Driving, Online = Parking, Charging, Sleeping)  for the EVENTS.  I'm not entirely clear on how the STATES data is generated, but could it be JOINED to provide the duration of each STATE/EVENT?  As I had requested something similar, in another thread for a new dashboard,  would it be possible to include the **START / END Rated Range** as well please?

![image](https://user-images.githubusercontent.com/30555695/120880791-9a02d380-c589-11eb-988d-f5583df3b220.png)

Thanks so much!

**Reply by jun3280net** [reply]
Figured some of this out by poking around inside Grafana.  I see how each of the Actions are captured in each section prior to the Unions.  I wish I had a SQL editor to do this, but I manually changed the SQL in Grafana and updated some of the Field Formatting to produce this update - which includes just the Start Range and End Range.  I enabled the ability to Filter on the Headers.  Is there a way to Word Wrap the Headers?  

![image](https://user-images.githubusercontent.com/30555695/120884478-b1998680-c5a0-11eb-992b-3e44820d2263.png)



Thanks again for this cool report!  This one table covers most of the relevant information I'm interested in.

I've attached my edits below if anyone wants to build on it.
[20210605_timeline.zip](https://github.com/adriankumpf/teslamate/files/6602197/20210605_timeline.zip)

**Reply by jun3280net** [reply]
I noticed @Dulanic was using PG Admin to query against the Teslamate db.  @adriankumpf is it possible to query the db from outside the docker-compose?

I installed PG Admin on Home Assistant which is on the same local machine as Teslamate.

![image](https://user-images.githubusercontent.com/30555695/120899590-bdf80080-c5ed-11eb-8483-7976789451ac.png)

**Reply by papakwstas** [reply]
Hi used the one with the map but there is no car map data showing. Any ideas?

**Comment by DrMichael** [comment]
Not sure about HomeAssistant, but when you expose the database port 5432 e.g. with `127.0.0.1:5432:5432` or `5432:5432`
in the docker-compose.yml, you should be able to access it via phppgadmin.

**Reply by jun3280net** [reply]
@DrMichael .... below is what I've adjusted in my docker-compose.  Unfortunately, I still can not access from PG Admin in Home Assistant.  I'm wondering do I need to port forward my router to 5432?  Perhaps I'm not entering the password properly.... is there any way to confirm?  I may extract PG Admin from  @Dulanic's docker-compose file.

![image](https://user-images.githubusercontent.com/30555695/120902040-5183fe00-c5fb-11eb-8ea5-95530322ef84.png)

**Comment by Dulanic** [comment]
@jun3280net  as long as you expose your port, you can connect to it from anything. I have a lot of things in docker, this is mine at https://github.com/Dulanic/docker-compose/blob/main/docker-compose.yml. I just connect to it as database:5432. I also connect to it from my desktop using HeidiSQL and also use a linked server from my SQL server using psqlodbc. 

I do use HA too and I connect to it /w this 

```
#Postgres SQL
recorder:
  db_url: !secret database
  exclude: 
    entities:
      - proximity.home_tesla
      - sensor.memory_free
      - sensor.network_in_enp2s0
      - sensor.network_out_enp2s0
      - sensor.tesla_speed_mi
      - sensor.load_5m
      - automation.update_tesla_location_as_mqtt_location_updates
      - sensor.tesla_longitude
      - sensor.tesla_latitude
      - media_player.shield
```

I also track my db size /w HA using this sensor:
```
 - platform: sql
   db_url: !secret database
   queries:
    - name: DB size
      query: "SELECT (pg_database_size('homeassistant')/1024/1024) as db_size;"
      column: "db_size"
      unit_of_measurement: MB  
```

**Reply by jun3280net** [reply]
@Dulanic  tried adding your PGADMIN docker compose snippet to my docker compose.  It half works - but I'm getting a permissions error.  Any idea how to solve this?  Thanks

![image](https://user-images.githubusercontent.com/30555695/120902852-ee489a80-c5ff-11eb-8b2b-c941c229bed9.png)

**Reply by Dulanic** [reply]
Sorry put wrong thing. pgadmin can be fun.... the folder must be owned by 5050:5050. You can see the info here...https://www.pgadmin.org/docs/pgadmin4/latest/container_deployment.html#mapped-files-and-directories

**Reply by jun3280net** [reply]
> Sorry put wrong thing. pgadmin can be fun.... the folder must be owned by 5050:5050. You can see the info here...https://www.pgadmin.org/docs/pgadmin4/latest/container_deployment.html#mapped-files-and-directories

Thanks @Dulanic.  I eventually got it working in both Home Assistant and via docker compose.  My main problem with the original Home Assistant, is that it didn't like 127.0.0.1.  However, if I put the IP of the machine, it worked fine.  Conversely, using Docker-Compose , I ended up following / adapting this guide :  [https://linuxhint.com/postgresql_docker/](https://linuxhint.com/postgresql_docker/) - which uses a LINK to make the connection.  Thanks for your help.  

PGadmin definitely makes the information from Teslamate more accessible, and meaningful.  Time to play around with the queries now :)

**Reply by Dulanic** [reply]
@jun3280net So docker has its own DNS. As long as they are on the same docker network, which if you didnt set a network they are, you can refer to other containers via name. You can never use 127.0.0.1 as they would only refer to that specific container. The only downside of links is they are a bandaid for any real issues and are being deprecated. You can use 
```
dockercontainername:port
hostip:port
or if port forwarded
extrnalip:port.
```
Anyways example, this is my pgadmin connection and this will be the fastest:
![image](https://user-images.githubusercontent.com/2948206/120928826-fe1db880-c6ab-11eb-9857-5de4f09d845e.png)

Example this is my nginx reverse proxy conf for teslamate:

```
        include /config/nginx/proxy.conf;
        resolver 127.0.0.11 valid=30s;
        proxy_pass http://teslamate:4000;
```

And yes, it makes things much easier to play with the queries, I also use pghero for information also but thats prob more than you need.

**Reply by jun3280net** [reply]
@Dulanic - thanks so much for taking the time to go through that and explain the setup.  Both HA and Docker setups were the the web based PG Admin - which times out after a few minutes.  I ended up installing PG Admin on my local Windows laptop.  Works much better now.  Getting into the queries now - but it turns out - the queries extracted from Grafana - has Grafana specific logic/functions.

For the Drives section, it appears the purpose of the LEFT JOIN with the Positions table is to support the START_PATH and END_PATH.  What is the intent of those two fields - as mine shows up like gibberish in Grafana.

On that note, I'm digging into what's going on with a recent DRIVE, PARK, CHARGE.  The CHARGE info is correct (START and END Range are correct.  The Drive End Range is wrong.  The Park Start and End Range are wrong.  

![image](https://user-images.githubusercontent.com/30555695/120933195-4215bb00-c6b6-11eb-9246-c92dc08f5821.png)

Will see what I can figure out :)

**Reply by Dulanic** [reply]
Just edit the panel, go to the query, click query inspector and it will show you the converted SQL. Also, not sure which verson you're reviewing. Mine doesnt look like what you posted /w the most recent one @DrMichael  posted, or maybe i missed one.

![image](https://user-images.githubusercontent.com/2948206/120937629-2de0b680-c6d4-11eb-8d0d-bdc92340aca4.png)

**Reply by jun3280net** [reply]
@Dulanic  thanks again for your tips.  I was always just working off the SQL after Editing the Panel.  The picture I posted was my edit of @DrMichael's last post.  I modified it a bit to add the Start and End Range.  Then I played around with the features in Grafana to see what else was possible.  It's more experimental in nature, but it delivers the information I need for now.

**Reply by jun3280net** [reply]
I'm not entirely clear how the Parking Section works, but I adjusted the End Rng, Rng Diff, and kWh by removing this from the formulas "**+ (LEAD(d.start_km) over w - d.end_km)**".  I believe that fixes the bulk of the problems I previously noted.  The End Rng never should have gotten to 434km.

![image](https://user-images.githubusercontent.com/30555695/121138318-cc643900-c7f4-11eb-81a3-268a3a05cb2d.png)

The End Rng for the Drive (2021-06-03 21:01:03) preceding the affected Park and Charge (2021-06-04 21:35:38), intuitively should have ended at 92km.  

Checking the Rated Range in Home Assistant history shows I'm missing a days worth of data here for (2021-06-04), which is why this looks weird.  I think this might have had something to do with the Wifi connection in the car - as my car was showing up as Offline that day.  I've since shut off Wifi - and now the car shows up as Sleeping when Parked for an extended period of time.   

![image](https://user-images.githubusercontent.com/30555695/121201450-c5a7e700-c831-11eb-8a10-829e266ba3fc.png)

**Reply by jun3280net** [reply]
Latest Dashboard with above mentioned formula adjustment, if anyone is interested.

![image](https://user-images.githubusercontent.com/30555695/121213394-68b12e80-c83b-11eb-8eab-3ebd4f08343f.png)


[Timeline-1623165986391.zip](https://github.com/adriankumpf/teslamate/files/6617808/Timeline-1623165986391.zip)

**Reply by DrMichael** [reply]
I would prefer not to double too much. This is a mixture of Overview, States and Timeline... (but also looks nice.)

> For the Drives section, it appears the purpose of the LEFT JOIN with the Positions table is to support the START_PATH and END_PATH. What is the intent of those two fields - as mine shows up like gibberish in Grafana.

Oh, those are indeed not for human consumption. They are used in the links...

> On that note, I'm digging into what's going on with a recent DRIVE, PARK, CHARGE. The CHARGE info is correct (START and END Range are correct. The Drive End Range is wrong. The Park Start and End Range are wrong.

Well in general it should not be necessary to have both Start Rng and End Rng, because Start Rng should equal End Rng of the previous drive. Which of course could be slightly different for Charging...

**Comment by DrMichael** [comment]
Ok, I worked a bit on that. So, now Odometer, Range, SoC are (hopefully) all consistent at the end of the respective period.

Though I like colours, but it is more or less personal flavour. (And for km and mi different...) SO, I left that out.
[timeline.zip](https://github.com/adriankumpf/teslamate/files/6622940/timeline.zip)

![image](https://user-images.githubusercontent.com/8087682/121336730-17c53700-c91c-11eb-9eba-adf6a5b3720a.png)

**Reply by jun3280net** [reply]
Thank you @DrMichael for your responses and continued updates!

**Comment by DrMichael** [comment]
I created a PR https://github.com/adriankumpf/teslamate/pull/1708

**Comment by enigma2k** [comment]
Thanks for the timeline implementation! I just got a request regarding the duration format:

Would it be possible to write instead of e.g.:
1.50 hour --> 1h 30m
1.5 day --> 1d 12h

this makes it much easier to read as you don't have to calculate yourself when you got more complex numbers like "1.17 hour".

**Reply by jun3280net** [reply]
> Thanks for the timeline implementation! I just got a request regarding the duration format:
> 
> Would it be possible to write instead of e.g.:
> 1.50 hour --> 1h 30m
> 1.5 day --> 1d 12h
> 
> this makes it much easier to read as you don't have to calculate yourself when you got more complex numbers like "1.17 hour".

@enigma2k  - you can change the format by Editing the Panel.  Then goto Overrides -> Override 9 (Duration).  Change Standard Options > Unit > Time > duration (hh:mm:ss)

![image](https://user-images.githubusercontent.com/30555695/122520669-23c29000-cfd1-11eb-8211-d0b8babf17a6.png)

**Reply by enigma2k** [reply]
@jun3280net 
 thank you! Unfortunately, it behaves a bit strangely.

e.g. when using the hh:mm:ss format it puts the minutes as seconds and hours as minutes. So 1.5 hour from the default setting changes to 00:01:30 and 53min changes to 00:00:53

I guess this happens because the seconds are missing in this log. is there a way to just remove the seconds from the template so I'm just getting 00:00?

and will all these settings remain after an update of teslamate?

**Reply by jun3280net** [reply]
> @jun3280net
> thank you! Unfortunately, it behaves a bit strangely.
> 
> e.g. when using the hh:mm:ss format it puts the minutes as seconds and hours as minutes. So 1.5 hour from the default setting changes to 00:01:30 and 53min changes to 00:00:53
> 
> I guess this happens because the seconds are missing in this log. is there a way to just remove the seconds from the template so I'm just getting 00:00?
> 
> and will all these settings remain after an update of teslamate?

hi @enigma2k.  I'm new to Grafana, so most of what I convey is just from playing around with the functions.  The changes are saved within the Dashboard and should be unaffected by Teslamate, unless they change the data structure in the database.  You can always SAVE another copy of the Dashboard with a different name as well if you're concerned.  Or if something changes with the data, you would need to edit the SQL.

On that note - I think Grafana might be mislabeling things.  Based on what you've conveyed, I think what is being output is **DD:HH:MM** which jives with what you've posted as an example.  

Here's what I see : 
![image](https://user-images.githubusercontent.com/30555695/122592532-9f4c2d80-d021-11eb-911e-76e5ac2301bc.png)

![image](https://user-images.githubusercontent.com/30555695/122592628-c60a6400-d021-11eb-9c2b-6ba37160af6c.png)

**Comment by jun3280net** [comment]
Updated my Grafana panel which is probably busier (and more colorful) than most prefer.  This serves as a landing page for me to quickly observe my vehicle's performance and current metrics.   There may be some useful pieces you may want to extract or expand upon for your own use.  Delete and modify as needed.  

While developing the SQL query, I restructured using WITH syntax.  I developed everything in my preferred units - KM and rated range.  However, I've added back the [[preferred_range]] syntax for those that want to dynamically switch between IDEAL and RATED.  The labels and units you'll need to switch if you use Miles.

1.  Updated Timeline Query to split out Charge Duration from the Park Duration.  This is a personal preference.  (eg Park and Charge used to start at the same time so there would be overlap).  Now it will show Park->Charge->Park.
2. I also removed MISSING action, as PARKING is the catch all for the data gaps between DRIVING, CHARGING, and UPDATING the way it's setup.
3. Added a graph to display how Range evolves through each of the Actions in the Timeline.

![image](https://user-images.githubusercontent.com/30555695/122603126-2654d200-d031-11eb-9d61-f939f56be8a5.png)


[Timeline_With-20210618.zip](https://github.com/adriankumpf/teslamate/files/6678700/Timeline_With-20210618.zip)

That said all the credit goes to @DrMichael for providing the foundation for me to learn how the data could be extracted and transformed.

**Reply by cwanja** [reply]
This is great, but I actually see this more as an overview dashboard as opposed to a timeline. Just given the amount of data on it. 

Not sure I follow what the colors in the timeline portion of the dashboard mean.

**Reply by jun3280net** [reply]
> This is great, but I actually see this more as an overview dashboard as opposed to a timeline. Just given the amount of data on it.
> 
> Not sure I follow what the colors in the timeline portion of the dashboard mean.

@cwanja  yup - its more of an Overview so I can just look at everything in one click.  @DrMichael's is the original Timeline view with more details exposed.   

If you delete all the other panels (aside from the Timeline), you get a more pure view as well.  The colors was just me playing with the Threshold feature in Grafana.  I didn't really apply it consistently between Range and Duration.  However, for Range - Red basically alerts me to Charge soon.  The one significant change I implemented was point 1 - which was splitting out Charge from Park data.  This was my personal preference.  The additional graph I added based on the Timeline data, was something I did for fun.

**Reply by jayali786** [reply]
How can I change C to F and also KM to Miles

Thank You, and also can we add map ?

**Comment by DrMichael** [comment]
I guess, we can close this. Timeline is there.
